File: year2021/day04/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static int[] NUMBERS;
    private static int[][][] BOARDS;

    public static void main(String[] args) { 
        try {
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        Set<Integer> marked = new HashSet<>();
        for (int num : NUMBERS) {
            marked.add(num);
            for (int[][] board : BOARDS) {
                if (containsMarkedRows(board, marked) || containsMarkedColumns(board, marked)) {
                    return num * calculateSum(board, marked);
                }
            }
        }
        return -1;
    }
    
    public static int part2() {
        Set<Integer> won = new HashSet<>();
        Set<Integer> marked = new HashSet<>();
        for (int num : NUMBERS) {
            marked.add(num);
            for (int i = 0; i < BOARDS.length; i++) {
                if (won.contains(i)) {
                    continue;
                }
                int[][] b = BOARDS[i];
                if (containsMarkedRows(b, marked) || containsMarkedColumns(b, marked)) {
                    won.add(i);
                    if (won.size() == BOARDS.length) {
                        return num * calculateSum(b, marked);
                    }
                }
            }
        }
        return -1;
    }

    public static void setup(String input) {
        NUMBERS = Arrays.stream(input.split("\n\n")[0].trim().split(","))
                .mapToInt(Integer::parseInt)
                .toArray();
        BOARDS = Arrays.stream(Arrays.copyOfRange(input.split("\n\n"), 1, input.split("\n\n").length))
                .map(board -> Arrays.stream(board.trim().split("\\n"))
                .map(row -> Arrays.stream(row.trim().split("\\s+"))
                .mapToInt(Integer::parseInt)
                .toArray())
                .toArray(int[][]::new))
                .toArray(int[][][]::new);   
    }

    private static boolean containsMarkedRows(int[][] board, Set<Integer> marked) {
        for (int[] row : board) {
            if (containsMarked(row, marked)) {
                return true;
            }
        }
        return false;
    }

    private static boolean containsMarkedColumns(int[][] board, Set<Integer> marked) {
        for (int columnIndex = 0; columnIndex < board[0].length; columnIndex++) {
            int[] column = new int[board.length];
            for (int rowIndex = 0; rowIndex < board.length; rowIndex++) {
                column[rowIndex] = board[rowIndex][columnIndex];
            }
            if (containsMarked(column, marked)) {
                return true;
            }
        }
        return false;
    }

    private static boolean containsMarked(int[] array, Set<Integer> marked) {
        for (int x : array) {
            if (!marked.contains(x)) {
                return false;
            }
        }
        return true;
    }

    private static int calculateSum(int[][] board, Set<Integer> marked) {
        int sum = 0;
        for (int[] row : board) {
            for (int x : row) {
                if (!marked.contains(x)) {
                    sum += x;
                }
            }
        }
        return sum;
    }
}

File: year2021/day04/python/solution.py
from itertools import chain

def main():
    print(part_1())
    print(part_2())


def part_1():
    marked = set()
    for num in numbers:
        marked.add(num)

        for board in boards:
            if any(all(x in marked for x in row) for row in board) or any(all(row[columnIndex] in marked for row in board) for columnIndex in range(len(board[0]))):
                return num * sum(x for x in chain.from_iterable(board) if x not in marked)
            

def part_2():
    won = set()
    marked = set()
    for num in numbers:
        marked.add(num)

        for i in range(len(boards)):
            if i in won:
                continue
            
            b = boards[i]
            
            if any(all(x in marked for x in row) for row in b) or any(all(row[columnIndex] in marked for row in b) for columnIndex in range(len(b[0]))):
                won.add(i)
                
                if len(won) == len(boards):
                    return num * sum(x for row in b for x in row if x not in marked)
                

def setup(path):
    global numbers
    global boards
    
    with open(path, 'r') as file:
        input = file.read().split('\n\n')
        numbers = [int(num) for num in input[0].split(',')]
        boards = [[[int(cell) for cell in row.split()] for row in board.split('\n')] for board in input[1:]]


if __name__ == '__main__':
    setup('../puzzle.txt')
    main()